### ISSUE TITLE
Bug: Return type of `tseslint.config()` is not compatible with eslint `Linter.FlatConfig[]`

### ISSUE THREAD

**Issue by latin-1** [typescript-eslint/typescript-eslint#9282]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-eslint

### Repro Code

```TypeScript
import type { Linter } from "eslint";
import tseslint from "typescript-eslint";

const config: Linter.FlatConfig[] = tsesling.config();
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

No type errors.

### Actual Result

With `"exactOptionalPropertyTypes": false`:

```
Type 'ConfigArray' is not assignable to type 'FlatConfig<RulesRecord>[]'.
  Type 'Config' is not assignable to type 'FlatConfig<RulesRecord>'.
    Types of property 'languageOptions' are incompatible.
      Type 'LanguageOptions | undefined' is not assignable to type '{ ecmaVersion?: 3 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 2015 | 2016 | 2017 | 2018 | 2019 | 2020 | 2021 | 2022 | 2023 | 2024 | "latest" | undefined; sourceType?: "script" | "module" | "commonjs" | undefined; globals?: Globals | undefined; parser?: FlatConfigParserModule | undefined; parserOptions?: Pars...'.
        Type 'LanguageOptions' is not assignable to type '{ ecmaVersion?: 3 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 2015 | 2016 | 2017 | 2018 | 2019 | 2020 | 2021 | 2022 | 2023 | 2024 | "latest" | undefined; sourceType?: "script" | "module" | "commonjs" | undefined; globals?: Globals | undefined; parser?: FlatConfigParserModule | undefined; parserOptions?: Pars...'.
          Types of property 'parser' are incompatible.
            Type 'LooseParserModule | undefined' is not assignable to type 'FlatConfigParserModule | undefined'.
              Type '{ meta?: { name?: string | undefined; version?: string | undefined; } | undefined; parseForESLint(text: string, options?: unknown): { ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }; }' is not assignable to type 'FlatConfigParserModule | undefined'.
                Type '{ meta?: { name?: string | undefined; version?: string | undefined; } | undefined; parseForESLint(text: string, options?: unknown): { ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }; }' is not assignable to type 'Omit<ParserModule, "parseForESLint"> & { parseForESLint(text: string, options?: any): Omit<ESLintParseResult, "ast" | "scopeManager"> & { ...; }; }'.
                  Type '{ meta?: { name?: string | undefined; version?: string | undefined; } | undefined; parseForESLint(text: string, options?: unknown): { ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }; }' is not assignable to type '{ parseForESLint(text: string, options?: any): Omit<ESLintParseResult, "ast" | "scopeManager"> & { ast: unknown; scopeManager?: unknown; }; }'.
                    The types returned by 'parseForESLint(...)' are incompatible between these types.
                      Type '{ ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }' is not assignable to type 'Omit<ESLintParseResult, "ast" | "scopeManager"> & { ast: unknown; scopeManager?: unknown; }'.
                        Type '{ ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }' is not assignable to type 'Omit<ESLintParseResult, "ast" | "scopeManager">'.
                          Types of property 'visitorKeys' are incompatible.
                            Type 'unknown' is not assignable to type 'VisitorKeys | undefined'.
```

With `"exactOptionalPropertyTypes": true`:
```
Type 'ConfigArray' is not assignable to type 'FlatConfig<RulesRecord>[]'.
  Type 'Config' is not assignable to type 'FlatConfig<RulesRecord>' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the types of the target's properties.
    The types of 'languageOptions.parser' are incompatible between these types.
      Type 'LooseParserModule' is not assignable to type 'FlatConfigParserModule' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the types of the target's properties.
        Type '{ meta?: { name?: string | undefined; version?: string | undefined; }; parseForESLint(text: string, options?: unknown): { ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }; }' is not assignable to type 'FlatConfigParserModule' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the types of the target's properties.
          Type '{ meta?: { name?: string | undefined; version?: string | undefined; }; parseForESLint(text: string, options?: unknown): { ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }; }' is not assignable to type 'Omit<ParserModule, "parseForESLint"> & { parseForESLint(text: string, options?: any): Omit<ESLintParseResult, "ast" | "scopeManager"> & { ...; }; }' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the types of the target's properties.
            Type '{ meta?: { name?: string | undefined; version?: string | undefined; }; parseForESLint(text: string, options?: unknown): { ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }; }' is not assignable to type '{ parseForESLint(text: string, options?: any): Omit<ESLintParseResult, "ast" | "scopeManager"> & { ast: unknown; scopeManager?: unknown; }; }'.
              The types returned by 'parseForESLint(...)' are incompatible between these types.
                Type '{ ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }' is not assignable to type 'Omit<ESLintParseResult, "ast" | "scopeManager"> & { ast: unknown; scopeManager?: unknown; }'.
                  Type '{ ast: unknown; services?: unknown; scopeManager?: unknown; visitorKeys?: unknown; }' is not assignable to type 'Omit<ESLintParseResult, "ast" | "scopeManager">'.
                    Types of property 'visitorKeys' are incompatible.
                      Type 'unknown' is not assignable to type 'VisitorKeys | undefined'.
```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `typescript-eslint`                    | `8.0.0-alpha.27` |
| `TypeScript`                           | `5.4.5` |
| `ESLint`                               | `9.4.0` |
| `@types/eslint`                        | `8.56.10` |
| `node`                                 | `X.Y.Z` |
